### ISSUE TITLE
Develop a "primary display" concept for Wayland

### ISSUE THREAD

**Issue by RPigott** [swaywm/sway#5758]
I thought of this while discussing #5647 but realized it was really a separate feature request.

The TL;DR is XRandR has a primary output concept and some apps, like bars, use that to decide where to be. i3bar is an example using the i3 syntax [output primary](https://i3wm.org/docs/userguide.html#_output_s) which is currently unsupported in sway.

I think that layer-shell clients should be able to request display on a primary output and have the compositor update their location when the primary changes. I also think that wlr-output-management should be updated to allow clients like kanshi to update the primary output in response to hotplug events. I'm not sure if a Wayland primary output could be made useful to X clients seeking a primary output.

The primary output I am proposing is different than the layer-shell default output. layer shell apps shouldn't default to the primary output unless they request it.

Do you know of any other X apps that make use of the primary output? Is it useful? Discuss.

**Comment by RPigott** [comment]
Related: i3/i3#4195

**Comment by emersion** [comment]
I'd rather make the concept of "primary output" a compositor policy, I think. And let users set layer-shell's default output to a specific output instead of the currently focused one.

**Comment by RPigott** [comment]
> And let users set layer-shell's default output to a specific output instead of the currently focused one.

I think we can allow that too. I also think it doesn't fulfill what a primary output concept does which is why I started a separate issue.

The core issue, I think, is that the compositor has no method to decide a default display that is suitable for _all_ layer-shell applications. The requirements of a bar or panel are different than a launcher or overlay. If I configure the layer-shell default output to always be some specific output so that my bar shows there, it will stop my launcher from appearing on the display I'm currently focusing.

I don't think the desired behavior can be reasonably achieved with any addition to sway. We cannot have per-application configuration like sway does with window criteria since layer-shell apps do not have unique app-ids. Even if they did I would prefer not to have to configure my applications that way. We also cannot have panels update smartly to hotplug events without merging the functionality of kanshi into sway, which is not ideal.

My proposal here is to have some way for clients to hint to the compositor how a default display should be selected _for them_. The compositor might reasonably ignore it. I chose "primary output" because it appears to fulfill that role in X, and is already being used in our X counterparts i3 and i3bar.

**Comment by emersion** [comment]
>We cannot have per-application configuration like sway does with window criteria since layer-shell apps do not have unique app-ids. 

layer-shell surfaces have a namespace, which I think suits well this use-case.

**Comment by RPigott** [comment]
> layer-shell surfaces have a namespace, which I think suits well this use-case.

OK, that seems good. I wasn't sure about using the namespace for this.

So in total what I think should be added is:

1. A sway command that allows users to configure the default output for layer-shell namespaces. Users should be able to specify a specific output, or set the default output as "primary" and possibly as "focused" to explicitly request the current default behavior.
2. A sway output command that sets which output is considered primary. Ideally, it would also advertise a matching XWAYLAND display as primary. Related discussion for mutter: https://gitlab.gnome.org/GNOME/mutter/-/issues/1407 , I would implement whatever method they go with.
3. When the primary output is changed, sway should move existing layer shell surfaces that have been assigned the primary output to the new one. This behavior is different than just assigning that layer-shell namespace directly to the output considered primary.
4. A swaybar option that will request the default layer shell output, so it can have behavior like i3bar's `output primary`.

**Comment by emersion** [comment]
TBH I think a command to assign layer surfaces to outputs (just like we can assign windows to workspaces) would be more generally useful than teaching Sway what a primary output is. Users to configure Xwayland's primary output with xrandr if they want to.

**Comment by zakk4223** [comment]
As requested (here: https://www.reddit.com/r/swaywm/comments/kpwre3/why_does_xrandr_list_my_monitors_as_xwayland8_and/gi1ph0f/)

I am setting my main monitor's XWAYLAND instance as primary, so that games pick up the proper resolution options for that monitor instead of using one of my smaller secondary displays.

However, if a monitor goes to sleep via DPMS, the XWAYLAND instance changes and has to be reset. It bumps the number of the instance every time. Right now my primary looks like this: 
XWAYLAND6 connected primary 3440x1440+1920+0 (normal left inverted right x axis 
y axis) 800mm x 330mm

Next time I sleep/wake it will be XWAYLAND7

I have an ugly hack to reset it, but it would be nice if it weren't something the user had to worry about.

**Comment by trougnouf** [comment]
> Do you know of any other X apps that make use of the primary output? Is it useful? Discuss.
Civilization VI, and I'm guessing many games will be whacky like that.

It used to work, but now it's limited to another monitor's resolution (1080p instead of 4K) so it won't let me increase the resolution and there is a mouse offset in fullscreen.

Now the only solution is to disable all of the other monitors when I'm playing.

It's certainly the game being so rigid but we can't fix every closed source game when they typically won't even patch anything critical thus requiring a workaround just to launch shortly after the libraries they used become outdated.

**Comment by zakk4223** [comment]
> > Do you know of any other X apps that make use of the primary output? Is it useful? Discuss.
> 
> Civilization VI, and I'm guessing many games will be whacky like that.
> 
> It used to work, but now it's limited to another monitor's resolution (1080p instead of 4K) so it won't let me increase the resolution and there is a mouse offset in fullscreen.
> 
> Now the only solution is to disable all of the other monitors when I'm playing.
> 
> It's certainly the game being so rigid but we can't fix every closed source game when they typically won't even patch anything critical thus requiring a workaround just to launch shortly after the libraries they used become outdated.

Use xrandr to set your 4K display as primary. 

As far as 'other X apps that make use of primary output?'. Almost all games that I've run have the issue. If you don't have a primary XWAYLAND instance, it uses whatever instance is at 0,0 (or leftmost? I'm not sure) to determine what video modes are available.
